--- a/firmware/fw_base.S
+++ b/firmware/fw_base.S
@@ -46,6 +46,16 @@
 	.globl _start
 	.globl _start_warm
 _start:
+	j	1f
+	/* TT: Place debug_descriptor pointer at a fixed address */
+	.align 7  # Align to 128-byte boundary (0x80)
+	.dword debug_descriptor - _start + FW_TEXT_START
+	/* TT: The zebu snapshot restore code needs 4K * 8 cores space here  */
+	/* each NOP is 2 bytes */
+	.rept 8 * (4096/2)
+	nop
+	.endr
+1:
 	/* Find preferred boot HART id */
 	MOV_3R	s0, a0, s1, a1, s2, a2
 	call	fw_boot_hart
@@ -133,7 +143,13 @@
 	/* Override previous arg1 */
 	lla	a1, fw_fdt_bin
 #endif
-
+#ifdef FW_JUMP_FDT_ADDR
+	li	a1, FW_JUMP_FDT_ADDR
+#elif defined(FW_JUMP_FDT_OFFSET)
+	lla	s5, _fw_start
+	li	a1, FW_JUMP_FDT_OFFSET
+	add	a1, s5, a1
+#endif
 	/*
 	 * Initialize platform
 	 * Note: The a0 to a4 registers passed to the
--- a/include/sbi/sbi_debug_descriptor.h
+++ b/include/sbi/sbi_debug_descriptor.h
@@ -0,0 +1,21 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2024 Tenstorrent, Inc.
+ *
+ * Authors:
+ *   Joel Smith <joelsmith@tenstorrent.com>
+ */
+
+#ifndef __SBI_DEBUG_DESCRIPTOR_H__
+#define __SBI_DEBUG_DESCRIPTOR_H__
+
+struct debug_descriptor {
+    u8 eye_catcher[8];  /* "OSBIdbug" */
+    u32 version;
+    u64 virtuart_base;
+};
+
+extern struct debug_descriptor debug_descriptor;
+
+#endif
--- a/include/sbi_utils/serial/virtual-uart.h
+++ b/include/sbi_utils/serial/virtual-uart.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2024 Tenstorrent Inc.
+ *
+ * Authors:
+ *   Joel Smith <joelsmith@tenstorrent.com>
+ */
+
+#ifndef __SERIAL_VIRTUAL_UART_H__
+#define __SERIAL_VIRTUAL_UART_H__
+
+#include <sbi/sbi_types.h>
+
+int virtual_uart_init();
+
+#endif
--- a/lib/sbi/sbi_init.c
+++ b/lib/sbi/sbi_init.c
@@ -33,6 +33,7 @@
 #include <sbi/sbi_tlb.h>
 #include <sbi/sbi_version.h>
 #include <sbi/sbi_unit_test.h>
+#include <sbi/sbi_debug_descriptor.h>
 
 #define BANNER                                              \
 	"   ____                    _____ ____ _____\n"     \
@@ -44,6 +45,12 @@
 	"        | |\n"                                     \
 	"        |_|\n\n"
 
+struct debug_descriptor debug_descriptor __attribute__((section(".data"))) = {
+	.eye_catcher = "OSBIdbug",
+	.version = 1,
+	.virtuart_base = -1ULL,
+};
+
 static void sbi_boot_print_banner(struct sbi_scratch *scratch)
 {
 	if (scratch->options & SBI_SCRATCH_NO_BOOT_PRINTS)
--- a/lib/utils/serial/Kconfig
+++ b/lib/utils/serial/Kconfig
@@ -92,4 +92,8 @@
 	bool "Semihosting support"
 	default n
 
+config SERIAL_VIRTUAL_UART
+	bool "Virtual UART"
+	default y
+
 endmenu
--- a/lib/utils/serial/objects.mk
+++ b/lib/utils/serial/objects.mk
@@ -46,3 +46,4 @@
 libsbiutils-objs-$(CONFIG_SERIAL_UART8250) += serial/uart8250.o
 libsbiutils-objs-$(CONFIG_SERIAL_XILINX_UARTLITE) += serial/xlnx-uartlite.o
 libsbiutils-objs-$(CONFIG_SERIAL_SEMIHOSTING) += serial/semihosting.o
+libsbiutils-objs-$(CONFIG_SERIAL_VIRTUAL_UART) += serial/virtual-uart.o
--- a/lib/utils/serial/virtual-uart.c
+++ b/lib/utils/serial/virtual-uart.c
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2024 Tenstorrent Inc.
+ *
+ * Authors:
+ *   Joel Smith <joelsmith@tenstorrent.com>
+ */
+
+#include <sbi/riscv_asm.h>
+#include <sbi/riscv_io.h>
+#include <sbi/sbi_console.h>
+#include <sbi/sbi_domain.h>
+#include <sbi/sbi_debug_descriptor.h>
+
+static const u64 VIRTUAL_UART_MAGIC = 0x5649525455415254ULL;
+
+#define BUFFER_SIZE 0x1000
+struct __attribute__((packed, aligned(4))) virtual_uart_queues {
+	volatile le64_t magic;
+	volatile char tx_buf[BUFFER_SIZE];
+	volatile char rx_buf[BUFFER_SIZE];
+	volatile le32_t tx_head;
+	volatile le32_t tx_tail;
+	volatile le32_t rx_head;
+	volatile le32_t rx_tail;
+};
+
+static struct virtual_uart_queues virtual_uart;
+
+static inline u32 read_ptr(volatile u32 *ptr)
+{
+	u32 value = *ptr;
+	smp_rmb();
+	return value;
+}
+
+static inline void write_ptr(volatile u32 *ptr, u32 value)
+{
+	smp_wmb();
+	*ptr = value;
+}
+
+static void virtual_uart_putc(char ch)
+{
+	volatile struct virtual_uart_queues *uart = &virtual_uart;
+	u32 head = read_ptr(&uart->tx_head);
+	u32 tail = read_ptr(&uart->tx_tail);
+	u32 next_head = (head + 1) % BUFFER_SIZE;
+
+	/* wait for the buffer to have space */
+	while (next_head == tail) {
+		tail = read_ptr(&uart->tx_tail);
+	}
+
+	if (next_head != tail) {
+		uart->tx_buf[uart->tx_head] = ch;
+		write_ptr(&uart->tx_head, next_head);
+	}
+}
+
+static int virtual_uart_getc(void)
+{
+	volatile struct virtual_uart_queues *uart = &virtual_uart;
+	u32 head = read_ptr(&uart->rx_head);
+	u32 tail = read_ptr(&uart->rx_tail);
+
+	if (head != tail) {
+		int ch = uart->rx_buf[tail];
+		smp_mb();
+		write_ptr(&uart->rx_tail, (tail + 1) % BUFFER_SIZE);
+		return ch;
+	}
+
+	return -1;
+}
+
+static struct sbi_console_device virtual_uart_console = {
+	.name = "virtual_uart",
+	.console_putc = virtual_uart_putc,
+	.console_getc = virtual_uart_getc
+};
+
+int virtual_uart_init()
+{
+	volatile struct virtual_uart_queues *uart = &virtual_uart;
+
+	uart->magic = VIRTUAL_UART_MAGIC;
+	uart->tx_head = 0;
+	uart->tx_tail = 0;
+	uart->rx_head = 0;
+	uart->rx_tail = 0;
+
+	debug_descriptor.virtuart_base = (u64)&virtual_uart;
+	sbi_console_set_device(&virtual_uart_console);
+
+	return 0;
+}
--- a/platform/generic/platform.c
+++ b/platform/generic/platform.c
@@ -11,6 +11,7 @@
 #include <platform_override.h>
 #include <sbi/riscv_asm.h>
 #include <sbi/sbi_bitops.h>
+#include <sbi/sbi_console.h>
 #include <sbi/sbi_hartmask.h>
 #include <sbi/sbi_heap.h>
 #include <sbi/sbi_platform.h>
@@ -29,6 +30,7 @@
 #include <sbi_utils/serial/fdt_serial.h>
 #include <sbi_utils/serial/semihosting.h>
 #include <sbi_utils/timer/fdt_timer.h>
+#include <sbi_utils/serial/virtual-uart.h>
 
 /* List of platform override modules generated at compile time */
 extern const struct fdt_driver *const platform_override_modules[];
@@ -225,6 +227,8 @@
 			rc = semihosting_init();
 		else
 			rc = fdt_serial_init(fdt);
+		if (rc == SBI_ENODEV)
+			rc = virtual_uart_init();
 		if (rc)
 			return rc;
 
